doc/hardware/cache/guide.rst

```diff
@@ -0,0 +1,169 @@
+cache support.
+
+.. note::
+
+  The information here assumes that the architecture-specific MPU support is
+  enabled. See the architecture-specific documentation for details.
+
+.. note::
+
+  While cache coherence can be a concern for data shared between SMP cores, Zephyr
+  in general ensures that memory will be seen in a coherent state from multiple
+  cores. Most applications will only need to use the cache APIs for interaction
+  with external hardware like DMA controllers or foreign CPUs running a
+  different OS image. For more information on cache coherence between SMP cores,
+  see :kconfig:option:`CONFIG_KERNEL_COHERENCE`.
+
+When dealing with memory shared between a processor core and other bus masters,
+cache coherency needs to be considered. Typically processor caches exist as
+close to each processor core as possible to maximize performance gain. Because
+of this, data moved into and out of memory by DMA engines will be stale in the
+processor's cache, resulting in what appears to be corrupt data. If you are
+moving data using DMA and the processor doesn't see the data you expect, cache
+coherency may be the issue.
+
+There are multiple approaches to ensuring that the data seen by the processor
+core and peripherals is coherent. The simplest is just to disable caching, but
+this defeats the purpose of having a hardware cache in the first place and
+results in a significant performance hit. Many architectures provide methods for
+disabling caching for only a portion of memory. This can be useful when cache
+coherence is more important than performance, such as when using DMA with SPI.
+Finally, there is the option to flush or invalidate the cache for regions of
+memory at runtime.
+
+Globally Disabling the Data Cache
+---------------------------------
+
+As mentioned above, globally disabling data caching can have a significant
+performance impact but can be useful for debugging.
+
+Requirements:
+
+* :kconfig:option:`CONFIG_DCACHE`: DCACHE control enabled in Zephyr.
+
+* :kconfig:option:`CONFIG_CACHE_MANAGEMENT`: cache API enabled.
+
+* Call :c:func:`sys_cache_data_disable()` to globally disable the data cache.
+
+Disabling Caching for a Memory Region
+-------------------------------------
+
+Disabling only a portion of memory can be a good performance compromise if
+performance on the uncached memory is not critical to the application. This is a
+good option if the application requires many small unrelated buffers that are
+smaller than a cache line.
+
+Requirements:
+
+* :kconfig:option:`CONFIG_DCACHE`: DCACHE control enabled in Zephyr.
+
+* :kconfig:option:`CONFIG_MEM_ATTR`: enable the ``mem-attr`` library for
+  handling memory attributes in the device tree.
+
+* Annotate your device tree according to :ref:`mem_mgmt_api`.
+
+Assuming the MPU driver is enabled, it will configure the specified regions
+according to the memory attributes specified during kernel initialization. When
```

Comment:
There's an additional gotcha here in some architectures in which number of MPU regions might not be sufficient if used for other proposes such as userspace. I don't know if that's the right place to discuss this but globally this is another choice that you might have to do in your general application design.

Reply:
Great point. I've added a couple notes calling this out and pointed to documentation for other features that may also require the use of dedicated MPU regions. Take a look and let me know if my additions match what you're thinking.

Reply:
Perfect, thanks